sources/packages/backend/libs/integrations/src/services/student-loan-balance/student-loan-balance.service.ts

```diff
@@ -0,0 +1,79 @@
+import { Injectable } from "@nestjs/common";
+import { SystemUsersService } from "@sims/services";
+import { StudentLoanBalance } from "@sims/sims-db";
+import { EntityManager } from "typeorm";
+
+@Injectable()
+export class StudentLoanBalanceService {
+  constructor(private readonly systemUserService: SystemUsersService) {}
+
+  /**
+   * Most recent balance present on database.
+   * If no records are present, returns undefined.
+   * @param referenceBalanceDate used as a reference to get the last balance before this date.
+   * @param entityManager used to get the data in and share the database transaction.
+   * @returns the most updated balance date, if present, otherwise, undefined.
+   */
+  async getLastBalanceDate(
+    referenceBalanceDate: Date,
+    entityManager: EntityManager,
+  ): Promise<Date | undefined> {
+    const maxBalanceDateAlias = "maxBalanceDate";
+    const lastBalanceDate = await entityManager
+      .getRepository(StudentLoanBalance)
+      .createQueryBuilder("studentLoanBalance")
+      .select("MAX(studentLoanBalance.balanceDate)", maxBalanceDateAlias)
+      .where("studentLoanBalance.balanceDate < :referenceBalanceDate", {
+        referenceBalanceDate,
+      })
+      .getRawOne();
+    return lastBalanceDate[maxBalanceDateAlias] ?? undefined;
+  }
+
+  /**
+   * After new student balance records are added to the database, checks if some students
+   * are no longer present in the statement which would indicate they no longer have
+   * a balance and a zero balance record will be inserted.
+   * @param previousBalanceDate most recent inserted balance date before the {@link currentBalanceDate}.
+   * @param currentBalanceDate date of the balance inserted into database that need
+   * to be checked if zero balance records are needed.
+   * @param entityManager used to get the data in and share the database transaction.
+   */
+  async insertZeroBalanceRecords(
+    previousBalanceDate: Date,
```

Comment:
What about deriving the previous balance date from the DB(2nd largest) instead of having the processing service to send it  here?

Reply:
The first version of that command would not allow it. With the current code, it would be possible.
The advantage was to avoid calling a second command when there were no records but it can also happen from the same method.
I will make the change.

Reply:
Refactored for now to have `getLastBalanceDate` as a private method. Only the `insertZeroBalanceRecords` will be called externally.